parsers/ttcn.c

```diff
@@ -0,0 +1,868 @@
+/*
+*   Copyright (c) 2009, Dmitry Peskov (dmitrypeskov@users.sourceforge.net)
+*
+*   This source code is released for free distribution under the terms of the
+*   GNU General Public License version 2 or (at your option) any later version.
+*
+*   File: ttcn.c
+*   Description: Adds TTCN-3 support to Exuberant ctags
+*   Version: 0.4
+*   Author: Dmitry Peskov (dmitrypeskov@users.sourceforge.net)
+*   Based on ETSI ES 201 873-1 V3.4.1 appendix A
+*   Version history:
+*   0.1     Initial version
+*   0.2     Enhancement: faster binary search in keyword list
+*           Fix: skip tag generation from import definitions (import/except lists)
+*   0.3     Fix: expression parser no longer relies on optional semicolon
+*           after a const/var/timer/modulepar definition to find where the initializer
+*           actually ends
+*           Fix: handle multiple module parameters not separated by semicolons
+*   0.31    Fix: lexer bug, missing character after a charstring close-quote
+*   0.4     Enhancement: tags for record/set/union members, enum values, port instances
+*           within components
+*/
+
+/* INCLUDE FILES */
+
+#include "general.h"    /* always include first */
+#include <string.h>     /* to declare strxxx() functions */
+#include <ctype.h>      /* to define isxxx() macros */
+#include "parse.h"      /* always include */
+#include "read.h"       /* to define file fileReadLine() */
+#include "routines.h"
+
+/*    DATA    */
+
+/* Tag kinds generated by parser */
+typedef enum {
+    K_MODULE,
+    K_TYPE,
+    K_CONST,
+    K_TEMPLATE,
+    K_FUNCTION,
+    K_SIGNATURE,
+    K_TESTCASE,
+    K_ALTSTEP,
+    K_GROUP,
+    K_MODULEPAR,
+    K_VAR,
+    K_TIMER,
+    K_PORT,
+    K_MEMBER,
+    K_ENUM,
+    K_NONE  /* No tag */
+} ttcnKind_t;
+static kindOption ttcnKinds [] = {

```

Comment:
The kind definitions are the heart of ctags.
So let's preview carefully.
Everything other than them can be fiexed later.